src/bin/mv.rs

```
@@ -1,16 +1,404 @@
+struct Arguments {
+    sources: Vec<PathBuf>,
+    target:  PathBuf,
+    flags:   Flags
+}
+
+/// Stores the state of each flag.
+struct Flags {
+    interactive: bool,
+    noclobber:   bool,
+    verbose:     bool,
+}
 
 fn main() {
-    let mut stderr = stderr();
-    let ref src = env::args().nth(1).fail("no source argument.", &mut stderr);
-    let ref dst = env::args().nth(2).fail("no destination argument.", &mut stderr);
+    let stderr = &mut io::stderr();
+    let stdout = io::stdout();
+    let stdout = &mut stdout.lock();
+    let arguments = env::args().skip(1).collect::<Vec<String>>();
+    mv(check_arguments(&arguments, stdout, stderr), stdout, stderr);
+}
+
+/// Take a list of arguments and attempt to move each source argument to their respective destination.
+fn mv(arguments: Arguments, stdout: &mut StdoutLock, stderr: &mut Stderr) {
+    for source in arguments.sources {
+        // Metadata from the source and target are required to determine both are on the same device.
+        let target_metadata = get_target_metadata(&arguments.target, &source, stderr);
+        let target = get_target_path(&arguments.target, &target_metadata, &source, stderr);
+        let source_metadata = match get_source_metadata(&source, stderr) {
+            Some(metadata) => metadata,
+            None => continue // We will skip this source because there was an error.
+        };
+
+        // Move the source file or directory to the target path.
+        // If the source and target are on the same device, rename the source.
+        // If they are on different devices, copy the file or directory.
+        if source_metadata.dev() == target_metadata.dev() {
+            match fs::rename(&source, &target) {
+                Ok(_) => if arguments.flags.verbose { verbose_print(&source, &target, stdout); },
+                Err(message) => {
+                    let _ = stderr.write(b"cannot rename '");
```

No, because letting this throw an error on not being able to print an error will not mean unbounded recursion. It will fail silently if the error about the error can't be printed, which is exactly how it should be.